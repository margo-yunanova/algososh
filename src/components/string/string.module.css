.form {
  display: flex;
  max-width: 567px;
  margin: auto;
  gap: 12px;
  padding-bottom: 120px;
}

.circle {
  display: flex;
  justify-content: center;
}
