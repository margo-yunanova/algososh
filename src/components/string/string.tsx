import React, { ChangeEvent, FormEventHandler, useState } from "react";
import { SolutionLayout } from "../ui/solution-layout/solution-layout";
import { Container } from "../content/container";
import { Input } from "../ui/input/input";
import { Button } from "../ui/button/button";
import styles from "./string.module.css";
import { Circle } from "../ui/circle/circle";
import { delay, swap } from "../../constants/utils";
import { ElementStates } from "../../types/element-states";
import { DELAY_IN_MS } from "../../constants/delays";

export const StringComponent: React.FC = () => {
  const [inputValue, setInputValue] = useState<string>("");
  const [{ firstIndex, secondIndex, data }, setDataForVisualization] =
    useState<{
      firstIndex?: number;
      secondIndex?: number;
      data: string[];
    }>({
      firstIndex: undefined,
      secondIndex: undefined,
      data: [],
    });
  const [hasStartedReverseInput, setStartReverseInput] = useState(false);

  const reverseInput = async (input: string) => {
    const array = input.split("");
    let lastIndex = array.length - 1;
    for (let i = 0; i < array.length / 2; i++) {
      swap(array, i, lastIndex);
      setDataForVisualization({
        firstIndex: i,
        secondIndex: lastIndex,
        data: array,
      });
      await delay(DELAY_IN_MS);
      lastIndex--;
    }
  };

  const getCircleState = (
    index: number,
    firstIndex: number | undefined,
    secondIndex: number | undefined
  ): ElementStates | undefined => {
    if (firstIndex === undefined || secondIndex === undefined)
      return ElementStates.Default;
    if (hasStartedReverseInput) {
      if (index === firstIndex || index === secondIndex)
        return ElementStates.Changing;

      if (index < firstIndex || index > secondIndex)
        return ElementStates.Modified;
    } else {
      if (index <= firstIndex || index >= secondIndex)
        return ElementStates.Modified;
    }
  };

  const submitHandler: FormEventHandler = async (e) => {
    e.preventDefault();
    setDataForVisualization({
      firstIndex: undefined,
      secondIndex: undefined,
      data: inputValue.split(""),
    });
    setStartReverseInput(true);
    await delay(DELAY_IN_MS);
    await reverseInput(inputValue);
    setStartReverseInput(false);
  };

  return (
    <SolutionLayout title="Строка">
      <Container padding="stringPage">
        <form className={styles.form} onSubmit={submitHandler}>
          <Input
            maxLength={11}
            isLimitText={true}
            value={inputValue}
            name="input"
            onChange={(e: ChangeEvent<HTMLInputElement>) =>
              setInputValue(e.target.value)
            }
          />
          <Button
            text="Развернуть"
            type="submit"
            isLoader={hasStartedReverseInput}
            disabled={inputValue === ""}
          />
        </form>
        <div className={styles.circle}>
          {data.map((letter, index) => (
            <Circle
              letter={letter}
              key={index}
              state={getCircleState(index, firstIndex, secondIndex)}
            />
          ))}
        </div>
      </Container>
    </SolutionLayout>
  );
};
