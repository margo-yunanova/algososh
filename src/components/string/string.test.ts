import { reverseString } from "./string";
jest.setTimeout(20000);

describe("reverse string", () => {
  it("even string length", async () => {
    const mockCallback = jest.fn();
    await reverseString("привет", mockCallback);
    expect(mockCallback).toMatchSnapshot();
  });

  it("odd string length", async () => {
    const mockCallback = jest.fn();
    await reverseString("кузинатра", mockCallback);
    expect(mockCallback).toMatchSnapshot();
  });

  it("one char", async () => {
    const mockCallback = jest.fn();
    await reverseString("я", mockCallback);
    expect(mockCallback).toMatchSnapshot();
  });

  it("empty string", async () => {
    const mockCallback = jest.fn();
    await reverseString("", mockCallback);
    expect(mockCallback).toMatchSnapshot();
  });
});
