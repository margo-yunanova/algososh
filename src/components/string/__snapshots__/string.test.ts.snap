// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reverse string empty string 1`] = `[MockFunction]`;

exports[`reverse string even string length 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Array [
          "т",
          "р",
          "и",
          "в",
          "е",
          "п",
        ],
        "firstIndex": 0,
        "secondIndex": 5,
      },
    ],
    Array [
      Object {
        "data": Array [
          "т",
          "е",
          "и",
          "в",
          "р",
          "п",
        ],
        "firstIndex": 1,
        "secondIndex": 4,
      },
    ],
    Array [
      Object {
        "data": Array [
          "т",
          "е",
          "в",
          "и",
          "р",
          "п",
        ],
        "firstIndex": 2,
        "secondIndex": 3,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`reverse string odd string length 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Array [
          "а",
          "у",
          "з",
          "и",
          "н",
          "а",
          "т",
          "р",
          "к",
        ],
        "firstIndex": 0,
        "secondIndex": 8,
      },
    ],
    Array [
      Object {
        "data": Array [
          "а",
          "р",
          "з",
          "и",
          "н",
          "а",
          "т",
          "у",
          "к",
        ],
        "firstIndex": 1,
        "secondIndex": 7,
      },
    ],
    Array [
      Object {
        "data": Array [
          "а",
          "р",
          "т",
          "и",
          "н",
          "а",
          "з",
          "у",
          "к",
        ],
        "firstIndex": 2,
        "secondIndex": 6,
      },
    ],
    Array [
      Object {
        "data": Array [
          "а",
          "р",
          "т",
          "а",
          "н",
          "и",
          "з",
          "у",
          "к",
        ],
        "firstIndex": 3,
        "secondIndex": 5,
      },
    ],
    Array [
      Object {
        "data": Array [
          "а",
          "р",
          "т",
          "а",
          "н",
          "и",
          "з",
          "у",
          "к",
        ],
        "firstIndex": 4,
        "secondIndex": 4,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`reverse string one char 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Array [
          "я",
        ],
        "firstIndex": 0,
        "secondIndex": 0,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
